app/assets/javascripts/discourse/app/lib/ember-events.js

```diff
@@ -0,0 +1,274 @@
+ * added via the `on` modifier. In particular, using `addEventListener`
+ * consistently everywhere ensures that event propagation works as expected
+ * between parent and child elements.
+ *
+ * @param {Application} app
+ */
+function rewireClassicComponentEvents(app) {
+  const allEvents = { ...EVENTS };
+
+  if (app.customEvents) {
+    for (const [event, methodName] of Object.entries(app.customEvents)) {
+      allEvents[event] = methodName;
+    }
+  }
+
+  const allEventMethods = {};
+  for (const [event, methodName] of Object.entries(allEvents)) {
+    allEventMethods[methodName] = event;
+  }
+
+  // Avoid Component.reopen to stop `ember.component.reopen` deprecation warning
+  EmberObject.reopen.call(Component, {
+    /**
+     * @param {string | typeof INTERNAL} name
+     * @param {unknown[]} args
+     */
+    trigger(name, ...args) {
+      if (name === INTERNAL) {
+        if (this.element) {
+          return this._super.call(this, ...args);
+        }
+      } else if (name.toLowerCase() in allEvents) {
+        return;
+      } else {
+        return this._super.call(this, name, ...args);
+      }
+    },
+
+    // eslint-disable-next-line ember/no-component-lifecycle-hooks
+    didInsertElement() {
+      this._super(...arguments);
+      setupComponentEventListeners(this, allEventMethods);
+    },
+
+    // eslint-disable-next-line ember/no-component-lifecycle-hooks
+    willDestroyElement() {
+      teardownComponentEventListeners(this);
+      this._super(...arguments);
+    },
+  });
+}
+
+/**
+ * Rewires the `action` modifier to use `addEventListener` directly instead of
+ * relying upon classic event delegation.
+ *
+ * This relies upon a deep override of Ember's rendering internals. If possible,
+ * consider eliminating usage of `action` as a modifier instead.
+ *
+ * @param {ApplicationInstance} appInstance
+ */
+function rewireActionModifier(appInstance) {
+  // This is a deep runtime override, since neither the runtime resolver nor the
+  // built-in `action` modifier seem to be available otherwise.
+  //
+  // TODO: Investigate if a cleaner override is possible.
+  const renderer = appInstance.lookup("renderer:-dom");
+  const lookupModifier = renderer._runtimeResolver.lookupModifier;
+  renderer._runtimeResolver.lookupModifier = (name, owner) => {
+    if (name === "action") {
+      return actionModifier;
+    } else {
+      return lookupModifier(name, owner);
+    }
+  };
+}
+
+function setupComponentEventListeners(component, allEventMethods) {
+  const proto = Object.getPrototypeOf(component);
+  let protoEvents = PROTO_EVENTS.get(proto);
+  const ownProps = Reflect.ownKeys(component);
+
+  // Memoize prototype event handlers at the prototype and add listeners
+  // to every instance.
+  if (!protoEvents) {
+    protoEvents = [];
+    PROTO_EVENTS.set(proto, protoEvents);
+
+    for (const method of Object.keys(allEventMethods)) {
+      if (component.has(method) && !ownProps.includes(method)) {
```

Comment:
Relying on `ownProps` here can cause the cached `protoEvents` to be incorrect. For example:

```javascript
// components/my-test-component.js

import Component from "@ember/component";

export default class extends Component {
  tagName = "button";
  click() {
    console.log("Default click handler");
  }
}
```

```hbs
<MyTestComponent @click={{this.someHandler}} />
<MyTestComponent />
```

The first component invocation correctly triggers `this.someHandler`.

However, the second component invocation does not have any event listeners registered because the cached `protoEvents` array is empty. The expected behaviour is for the prototype's `click()` method to be called.

Reply:
Good catch, @davidtaylorhq - I'll straighten this out asap.

Reply:
I've streamlined the implementation of `setupComponentEventListeners` in https://github.com/discourse/discourse/pull/18742/commits/3dd2861cfdb9eac3e40c392ee6209aab0366b87d

This commit removes memoization of events installed on the prototype, which gained us very little given the additional need to iterate through the same events on the component instance.

It also adds tests to ensure that events are configured whether they are set on the prototype or instance, and to ensure that the order of instantiation doesn't matter.

I think this is a win for code clarity and correctness, but let me know if there's a performance regression that I didn't anticipate.